### PR TITLE
refactor: use adaptive pooling

### PR DESCRIPTION
**Pull request by slettner**
Replaces the pooling_strategy which has to be set by the user with a fixed `AdaptiveAvgPooling2d` layer (=np.mean).
This takes the burden thinking about pooling strategies from the user by removing one parameter from the class.

### CODE REVIEW

File: encoder.py [hunk 1]
```diff
@@ -108,9 +100,9 @@ def _get_features(self, content):
         return self.model(content)
 
     def _get_pooling(self, feature_map: 'np.ndarray') -> 'np.ndarray':
```
Comment (JoanFM) [inline encoder.py]: This is not going to work on `GPU` right?

Also I am not sure why we need to take care of this? the first condition, does it ever happen? 

I think the pooling should be done in the torch tensor directly that may be in GPU as the layer

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: encoder.py (hunk 1 above)
Comment (slettner) [reply encoder.py]: I removed the if statement in https://github.com/jina-ai/executor-image-torch-encoder/pull/7/commits/4a4775e667e1787f03b34e74bb7659f3d65628c9

I would expect it to work on GPU now, but I can't test it locally. 
There is a unit tests for GPU usage, if you have one, could you try?

File: encoder.py [hunk 2]
```diff
@@ -145,11 +120,13 @@ def _compute_embeddings(self, docs_batch_generator: Iterable) -> None:
                 else:
                     images = np.stack(blob_batch)
                 tensor = torch.from_numpy(images).to(self.device)
-                features = self._get_features(tensor).detach()
-                features = self._get_pooling(features.cpu().numpy())
+                features = self._get_features(tensor)
```
Comment (JoanFM) [inline encoder.py]: let's merge this into a function `_get_embedding`

File: encoder.py [hunk 3]
```diff
@@ -65,22 +56,20 @@ def __init__(
...
         model = getattr(models, self.model_name)(pretrained=True)
 
         self.model = self._extract_feature_from_torch_module(model)
         self.model.to(torch.device(self.device))
-        if self.pool_strategy is not None:
-            self.pool_fn = getattr(np, self.pool_strategy)
+
+        self._pooling_layer = nn.AdaptiveAvgPool2d(1)
+        self._pooling_layer.to(torch.device(self.device))
+        self._pooling_function = lambda x: self._pooling_layer(x).squeeze(3).squeeze(2)
```
Comment (JoanFM) [inline encoder.py]: add this to the model instead of a new `lambda`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: encoder.py (hunk 2 above)
Comment (slettner) [reply encoder.py]: Done in https://github.com/jina-ai/executor-image-torch-encoder/pull/7/commits/9631c630ccfca61e2cbe9f4c5efb5ce1178de0c8

File: models.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    tuple(all_vgg_models[1:]): 'features',
+    tuple(all_squeezenet_models[1:]): 'features',
+    tuple(all_densenet_models[1:]): 'features',
+    tuple(all_mnasnet_models[1:]): 'layers',
+    tuple(all_mobilenet_models[1:]): 'features',
+}
+
+
+def is_model_supported(name: str) -> bool:
+    return sum([name in model_names for model_names in _all_supported_models_to_layer_mapping]) > 0
```
Comment (JoanFM) [inline models.py]: this is very weird, isn't there something like `any`? or have a flatten list of all?

File: setup.py [hunk 5]
```diff
@@ -12,7 +12,7 @@
         "Programming Language :: Python :: 3",
         "Operating System :: OS Independent",
     ],
-    py_modules=['jinahub.image.encoder.torch_encoder'],
+    py_modules=['jinahub.image.encoder.torch_encoder', 'jinahub.image.encoder.models'],
```
Comment (JoanFM) [inline setup.py:15]: why this?

File: tests/integration/test_flow_integration.py [hunk 6]
```diff
@@ -31,7 +31,7 @@ def test_no_batch(arr_in: np.ndarray):
     results_arr = DocumentArray(resp[0].data.docs)
     assert len(results_arr) == 1
     assert results_arr[0].embedding is not None
-    assert results_arr[0].embedding.shape == (1000,)
+    assert results_arr[0].embedding.shape == (1280, )
```
Comment (JoanFM) [inline tests/integration/test_flow_integration.py:34]: why has this changed?

File: torch_encoder.py [hunk 7]
```diff
@@ -66,22 +56,20 @@ def __init__(
...
         model = getattr(models, self.model_name)(pretrained=True)
 
         self.model = self._extract_feature_from_torch_module(model)
         self.model.to(torch.device(self.device))
-        if self.pool_strategy is not None:
-            self.pool_fn = getattr(np, self.pool_strategy)
+
+        self._pooling_layer = nn.AdaptiveAvgPool2d(1)
+        self._pooling_layer.to(torch.device(self.device))
+        self._pooling_function = lambda x: self._pooling_layer(x).squeeze(3).squeeze(2)
```
Comment (JoanFM) [inline torch_encoder.py]: I am not sure, why is `squeezing` necessary?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: torch_encoder.py (hunk 7 above)
Comment (slettner) [reply torch_encoder.py]: The `nn.AdaptiveAvgPool2d(1)`  returns a vector `B x D x 1 x 1`. It is to remove the one dimensions

File: tests/integration/test_flow_integration.py (hunk 6 above)
Comment (slettner) [reply tests/integration/test_flow_integration.py:34]: Before, this was using an incorrect resnet layer. 1000 is the classification layer.

File: torch_encoder.py [hunk 8]
```diff
@@ -66,22 +56,20 @@ def __init__(
...
-        self.pool_fn = getattr(np, self.pool_strategy)
 
         model = getattr(models, self.model_name)(pretrained=True)
 
         self.model = self._extract_feature_from_torch_module(model)
         self.model.to(torch.device(self.device))
-        if self.pool_strategy is not None:
-            self.pool_fn = getattr(np, self.pool_strategy)
+
+        self._pooling_layer = nn.AdaptiveAvgPool2d(1)
```
Comment (JoanFM) [inline torch_encoder.py]:
```suggestion
        self._pooling_layer = nn.AdaptiveAvgPool2d(output_size=(1, 1))
```
This should be without `squeeze`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: setup.py (hunk 5 above)
Comment (slettner) [reply setup.py:15]: We need this module installed. It contains relevant code for the executor.

File: models.py (hunk 4 above)
Comment (slettner) [reply models.py]: yep, thanks for the tip https://github.com/jina-ai/executor-image-torch-encoder/pull/7/commits/1ac07261ac5a2d1de8e657fdaaef10d027545a14

Comment (davidbp) [reply models.py]: This is the same as `return any([name in model_names for model_names in _all_supported_models_to_layer_mapping])`

File: models.py [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+        self._layer_name = _ModelCatalogue.get_layer_name(model_name)
+        self._model = getattr(models, model_name)(pretrained=True)
+
+        self.device = device
+
+        self._pooling_layer = nn.AdaptiveAvgPool2d(1)
+        self._pooling_layer.to(torch.device(self.device))
+
+    def _pooling_function(self, tensor_in: 'torch.Tensor') -> 'torch.Tensor':
+        return self._pooling_layer(tensor_in).squeeze(3).squeeze(2)
```
Comment (JoanFM) [inline models.py]: this squeezing is a smell

File: models.py [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    def __init__(self, model_name: str, device: Optional[str] = None):
+        if not device:
+            device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+        self._layer_name = _ModelCatalogue.get_layer_name(model_name)
+        self._model = getattr(models, model_name)(pretrained=True)
+
+        self.device = device
+
+        self._pooling_layer = nn.AdaptiveAvgPool2d(1)
```
Comment (JoanFM) [inline models.py]: try with  `nn.AdaptiveAvgPool2d(output_size=(1, 1))`

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: models.py (hunk 9 above)
Comment (slettner) [reply models.py]: Refactored in https://github.com/jina-ai/executor-image-torch-encoder/pull/7/commits/e20c9c7b4853241f9516cf32c40b7e7888bc3885

File: models.py (hunk 10 above)
Comment (slettner) [reply models.py]: Changed in https://github.com/jina-ai/executor-image-torch-encoder/pull/7/commits/e20c9c7b4853241f9516cf32c40b7e7888bc3885

File: torch_encoder.py [hunk 11]
```diff
@@ -124,19 +83,14 @@ def encode(self, docs: Optional[DocumentArray], parameters: Dict, **kwargs):
         :param kwargs: Additional key value arguments.
         """
         if docs:
-            docs_batch_generator = self._get_docs_batch_generator(docs, parameters)
+            docs_batch_generator = get_docs_batch_generator(
+                docs,
+                traversal_path=parameters.get('traversal_paths', self.default_traversal_path),
+                batch_size=parameters.get('batch_size', self.default_batch_size),
+                needs_attr='text'
```
Comment (JoanFM) [inline torch_encoder.py]: `text`?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: torch_encoder.py (hunk 11 above)
Comment (slettner) [reply torch_encoder.py]: Fixed in https://github.com/jina-ai/executor-image-torch-encoder/pull/7/commits/e8806b3f2804b23b1e466cabbd6a46f83a613766

**Review by JoanFM — APPROVED** [review]

**Comment by JoanFM** [comment]
LGTM! Good job @slettner !

**Pull request merged**